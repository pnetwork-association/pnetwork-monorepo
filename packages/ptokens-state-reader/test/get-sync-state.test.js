const constants = require('ptokens-constants')
const { db, logic } = require('ptokens-utils')
const { EventEmitter } = require('stream')
const {
  updateActorsPropagatedEventInStorage,
} = require('../lib/chains/evm/get-actors-for-current-epoch')
const {
  MEM_ACTORS,
  MEM_ACTORS_PROPAGATED,
  STATE_DB_ACTORS_PROPAGATED_KEY,
  STATE_DB_ACTORS_KEY,
  MEM_TIMESTAMP,
} = require('../lib/constants')
const guardianStatusSample = require('./mock/guardian-status-sample.json')
const actorsPropagatedSample = require('./chains/evm/mock/actors-propagated-sample.json')
const { getActorFromStorage } = require('../lib/get-actor-from-storage')
const { ipfs } = require('ptokens-utils')

describe('Get syncing status general tests', () => {
  let actorsStorage = null
  let actorsPropagatedStorage = null

  const uri = global.__MONGO_URI__
  const dbName = global.__MONGO_DB_NAME__

  const actorAddress = guardianStatusSample[constants.statusObject.KEY_SIGNER_ADDRESS]

  beforeAll(async () => {
    actorsStorage = await db.getCollection(uri, dbName, MEM_ACTORS)
    actorsPropagatedStorage = await db.getCollection(uri, dbName, MEM_ACTORS_PROPAGATED)
    await updateActorsPropagatedEventInStorage(actorsPropagatedStorage, actorsPropagatedSample)
  })

  beforeEach(async () => {
    jest.restoreAllMocks()
    await actorsStorage.deleteMany({})
  })

  afterAll(async () => {
    await db.closeConnection(uri)
  })

  it('Should insert a new actor after receiving the state for the first time', () => {
    const state = {
      [STATE_DB_ACTORS_KEY]: actorsStorage,
      [STATE_DB_ACTORS_PROPAGATED_KEY]: actorsPropagatedStorage,
      [constants.config.KEY_DB]: {
        [constants.config.KEY_URL]: global.__MONGO_URI__,
        [constants.config.KEY_NAME]: global.__MONGO_DB_NAME__,
      },
      [constants.config.KEY_PROTOCOLS]: [
        {
          [constants.config.KEY_TYPE]: 'ipfs',
          [constants.config.KEY_DATA]: {
            url: 'http://127.0.0.1:5001',
            topic: 'pnetwork-v3',
          },
        },
      ],
    }

    const eventEmitter = new EventEmitter()
    const subMock = jest.fn().mockResolvedValue(eventEmitter)
    jest.spyOn(ipfs, 'PubSub').mockImplementation(() => ({ sub: subMock }))

    const { getSyncStateAndUpdateTimestamps } = require('../lib/get-sync-state')
    getSyncStateAndUpdateTimestamps(state)

    return Promise.resolve(logic.sleepForXMilliseconds(100))
      .then(_ => eventEmitter.emit('message', JSON.stringify(guardianStatusSample)))
      .then(_ => logic.sleepForXMilliseconds(100))
      .then(_ => getActorFromStorage(actorsStorage, actorAddress))
      .then(_result =>
        expect(_result).toMatchSnapshot({
          [MEM_TIMESTAMP]: expect.any(Number),
        })
      )
  })
})
