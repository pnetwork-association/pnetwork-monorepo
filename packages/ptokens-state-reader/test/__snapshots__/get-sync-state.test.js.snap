// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get syncing status general tests Should insert a new actor after receiving the state for the first time 1`] = `
{
  "_id": "0xdb30d31ce9a22f36a44993b1079ad2d201e11788",
  "actor": "0xdb30d31ce9a22f36a44993b1079ad2d201e11788",
  "actorStatus": {
    "0x5aca268b": "Active",
    "0xb9286154": "Active",
    "0xf9b459a1": "Active",
  },
  "syncState": {
    "0x5aca268b": {
      "latestBlockHash": "0x02cda5236e6c43c32f2cf3427c86c48a48802c90b0f040dd0054e5f37ffbf7ed",
      "latestBlockNumber": 33483580,
      "latestBlockTimestamp": 1699958718,
    },
    "0xb9286154": {
      "latestBlockHash": "0xb19e0b149a9eaafa64234f94f1036e89b3150d29a004ab2d78a5723510a58e82",
      "latestBlockNumber": 10041664,
      "latestBlockTimestamp": 1699958712,
    },
    "0xf9b459a1": {
      "latestBlockHash": "0x84b021894c8456aff1054a41955554f862a5102b415ac14c5d2aa0e95c14cf66",
      "latestBlockNumber": 49930328,
      "latestBlockTimestamp": 1699958718,
    },
  },
  "timestamp": Any<Number>,
}
`;
